cpp/src/parquet/page_index.cc

```diff
@@ -0,0 +1,229 @@
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "parquet/page_index.h"
+#include "parquet/encoding.h"
+#include "parquet/exception.h"
+#include "parquet/schema.h"
+#include "parquet/statistics.h"
+#include "parquet/thrift_internal.h"
+
+#include "arrow/util/unreachable.h"
+
+#include <limits>
+#include <map>
+#include <numeric>
+
+namespace parquet {
+
+namespace {
+
+template <typename DType>
+void Decode(std::unique_ptr<typename EncodingTraits<DType>::Decoder>& decoder,
+            const std::string& input, std::vector<typename DType::c_type>& output,
+            size_t index) {
+  if (ARROW_PREDICT_FALSE(index >= output.size())) {
+    throw ParquetException("Index out of bound");
+  }
+
+  decoder->SetData(/*num_values=*/1, reinterpret_cast<const uint8_t*>(input.c_str()),
+                   static_cast<int>(input.size()));
+  decoder->Decode(&output[index], /*max_values=*/1);
+}
+
+template <>
+void Decode<BooleanType>(std::unique_ptr<BooleanDecoder>& decoder,
+                         const std::string& input, std::vector<bool>& output,
+                         size_t index) {
+  if (ARROW_PREDICT_FALSE(index >= output.size())) {
+    throw ParquetException("Index out of bound");
+  }
+
+  bool value;
+  decoder->SetData(/*num_values=*/1, reinterpret_cast<const uint8_t*>(input.c_str()),
+                   static_cast<int>(input.size()));
+  decoder->Decode(&value, /*max_values=*/1);
+  output[index] = value;
+}
+
+template <>
+void Decode<ByteArrayType>(std::unique_ptr<ByteArrayDecoder>&, const std::string& input,
+                           std::vector<ByteArray>& output, size_t index) {
+  if (ARROW_PREDICT_FALSE(index >= output.size())) {
+    throw ParquetException("Index out of bound");
+  }
+
+  if (ARROW_PREDICT_FALSE(input.size() >
+                          static_cast<size_t>(std::numeric_limits<uint32_t>::max()))) {
+    throw ParquetException("Invalid encoded byte array length");
+  }
+
+  auto& decoded = output.at(index);
+  decoded.len = static_cast<uint32_t>(input.size());
+  decoded.ptr = reinterpret_cast<const uint8_t*>(input.data());
+}
+
+template <typename DType>
+class TypedColumnIndexImpl : public TypedColumnIndex<DType> {
+ public:
+  using T = typename DType::c_type;
+
+  TypedColumnIndexImpl(const ColumnDescriptor& descr,
+                       const format::ColumnIndex& column_index)
+      : column_index_(column_index) {
+    // Make sure the number of pages is valid and it does not overflow to int32_t.
+    if (column_index_.null_pages.size() != column_index_.min_values.size() ||
+        column_index_.min_values.size() != column_index_.max_values.size() ||
+        ARROW_PREDICT_FALSE(column_index_.null_pages.size() >=
+                            static_cast<size_t>(std::numeric_limits<int32_t>::max()))) {
+      throw ParquetException("Invalid column index");
+    }
+
```

Comment:
Can we also check the size of `null_counts` if present?

Reply:
Fixed